Quaver.API/Maps/Structures/HitObjectInfo.cs

```diff
@@ -11,177 +11,184 @@
-        }
+    /// <summary>
+    ///     Returns if the object is allowed to be edited in lua scripts
+    /// </summary>
+    [YamlIgnore]
+    public bool IsEditableInLuaScript
+    {
+        get;
+        [MoonSharpVisible(false)] set;
+    }
 
-        /// <summary>
-        /// </summary>
-        /// <param name="time"></param>
-        /// <exception cref="InvalidOperationException"></exception>
-        public void SetStartTime(int time)
-        {
-            ThrowUneditableException();
-            StartTime = time;
-        }
+    /// <summary>
+    ///     Gets the timing point this object is in range of.
+    /// </summary>
+    /// <returns></returns>
+    public TimingPointInfo GetTimingPoint(List<TimingPointInfo> timingPoints) => timingPoints.AtTime(StartTime);
 
-        /// <summary>
-        /// </summary>
-        /// <exception cref="InvalidOperationException"></exception>
-        public void SetEndTime(int time)
-        {
-            ThrowUneditableException();
-            EndTime = time;
-        }
+    /// <summary>
+    /// </summary>
+    /// <param name="time"></param>
+    /// <exception cref="InvalidOperationException"></exception>
+    public void SetStartTime(int time)
+    {
+        ThrowUneditableException();
+        StartTime = time;
+    }
 
-        /// <summary>
-        /// </summary>
-        /// <param name="lane"></param>
-        public void SetLane(int lane)
-        {
-            ThrowUneditableException();
-            Lane = lane;
-        }
+    /// <summary>
+    /// </summary>
+    /// <exception cref="InvalidOperationException"></exception>
+    public void SetEndTime(int time)
+    {
+        ThrowUneditableException();
+        EndTime = time;
+    }
 
-        /// <summary>
-        /// </summary>
-        /// <param name="hitsounds"></param>
-        public void SetHitSounds(HitSounds hitsounds)
-        {
-            ThrowUneditableException();
-            HitSound = hitsounds;
-        }
+    /// <summary>
+    /// </summary>
+    /// <param name="lane"></param>
+    public void SetLane(int lane)
+    {
+        ThrowUneditableException();
+        Lane = lane;
+    }
 
-        /// <summary>
-        /// </summary>
-        /// <exception cref="InvalidOperationException"></exception>
-        private void ThrowUneditableException()
+    /// <summary>
+    /// </summary>
+    /// <param name="hitsounds"></param>
+    public void SetHitSounds(HitSounds hitsounds)
+    {
+        ThrowUneditableException();
+        HitSound = hitsounds;
+    }
+
+    /// <summary>
+    /// </summary>
+    /// <exception cref="InvalidOperationException"></exception>
+    private void ThrowUneditableException()
+    {
+        if (!IsEditableInLuaScript)
+            throw new InvalidOperationException("Value is not allowed to be edited in lua scripts.");
+    }
+
+    /// <inheritdoc />
+    public int CompareTo(HitObjectInfo other) => EditorLayer.CompareTo(other.EditorLayer);
```

Comment:
Don't understand why we're sorting by editor layer. Actually does it even make sense to define a default comparer for these objects? Now that there are dedicated sort functions for each collection of map objects, perhaps define a custom comparer function in them instead?

Reply:
Whoops, I don't know how I zoned out so hard as to do that. It's meant to be `StartTime`.

Reply:
I don't think a custom comparer is appropriate in this case because the game almost always sorts by `StartTime`, and it would involve having to change every call site to include the comparer.

Reply:
I thought about it more, and actually you're totally right. All usages in [this PR](https://github.com/Quaver/Quaver/pull/4105) do not use anything other than types that already implement `IStartTime`. So I've changed it to what you suggested.